src/main/flight/position_control.c

```diff
@@ -48,3 +58,50 @@ const pidCoefficient_t *getAltitudePidCoeffs(void)
 {
     return &altitudePidCoeffs;
 }
+
+void resetAltitudeControl (void) {
+    altitudeI = 0.0f;
+}
+
+void altitudeControl(float targetAltitudeCm, float taskIntervalS, float verticalVelocity, float targetAltitudeStep) {
+
+    float altitudeErrorCm = targetAltitudeCm - getAltitudeCm();
+    const float altitudeP = altitudeErrorCm * altitudePidCoeffs.Kp;
+
+    // reduce the iTerm gain for errors greater than 2m, otherwise it winds up too much
+    const float largeError = 200.0f; // 2m
+    const float itermRelax = (fabsf(altitudeErrorCm) < largeError) ? 1.0f : 0.1f;
+    altitudeI += altitudeErrorCm * altitudePidCoeffs.Ki * itermRelax * taskIntervalS;
+    const float iTermLimit = 200.0f;
+    altitudeI = constrainf(altitudeI, -iTermLimit, iTermLimit); 
+
+    const float altitudeD = verticalVelocity * altitudePidCoeffs.Kd;
+
+    const float altitudeF = targetAltitudeStep * altitudePidCoeffs.Kf;
+
+    const float throttleAdjustment = altitudeP + altitudeI - altitudeD + altitudeF;
+    const float tiltMultiplier = 2.0f - fmaxf(getCosTiltAngle(), 0.5f);
+    // 1 = flat, 1.24 at 40 degrees, max 1.5 around 60 degrees, the default limit of Angle Mode
+    // 2 - cos(x) is between 1/cos(x) and 1/sqrt(cos(x)) in this range
+
+    const float hoverThrottle = positionControlConfig()->hover_throttle - PWM_RANGE_MIN;
+
+    float newThrottle = (hoverThrottle + throttleAdjustment) * tiltMultiplier + PWM_RANGE_MIN;
+    DEBUG_SET(DEBUG_ALTITUDE_CONTROL, 2, lrintf(newThrottle)); // normal range 1000-2000 but is before constraint
+
+    newThrottle = constrainf(newThrottle, positionControlConfig()->alt_control_throttle_min, positionControlConfig()->alt_control_throttle_max);
+    newThrottle = scaleRangef(throttleOut, MAX(rxConfig()->mincheck, PWM_RANGE_MIN), PWM_RANGE_MAX, 0.0f, 1.0f);
+    throttleOut = constrainf(newThrottle, 0.0f, 1.0f);
+
+    DEBUG_SET(DEBUG_ALTITUDE_CONTROL, 0, lrintf(targetAltitudeCm));
+    DEBUG_SET(DEBUG_ALTITUDE_CONTROL, 3, lrintf(tiltMultiplier * 100));
+    DEBUG_SET(DEBUG_ALTITUDE_CONTROL, 4, lrintf(altitudeP));
+    DEBUG_SET(DEBUG_ALTITUDE_CONTROL, 5, lrintf(altitudeI));
+    DEBUG_SET(DEBUG_ALTITUDE_CONTROL, 6, lrintf(-altitudeD));
```

Comment:
```suggestion
    DEBUG_SET(DEBUG_ALTITUDE_CONTROL, 6, lrintf(altitudeD));
```

Reply:
Here I was intending to log 'negative altitudeD`, since the normal depiction of D in logs is that it carries the sign in which it is added to the PIDsum.
Because of the way we get velocity now from position.c, where positive means velocity upwards, the D correction must be subtracted from the PIDsum, and it is.
That means the D value itself is positive here, so the sign must be made negative in the log for it to reflect the D value used in the sum.